### PR TITLE
Scorched earth 2

### PR DESCRIPTION
**Pull request by wyfo**
Complete rewriting using only Rust code (and Python stub)

### CODE REVIEW

File: src/session.rs [hunk 1]
```diff
@@ -1,5 +1,5 @@
 //
-// Copyright (c) 2017, 2022 ZettaScale Technology Inc.
+// Copyright (c) 2024 ZettaScale Technology
```
Comment (evshary) [inline src/session.rs:2]: I think we don't need to update the year here.

File: src/queryable.rs [hunk 2]
```diff
@@ -1,5 +1,5 @@
 //
-// Copyright (c) 2017, 2022 ZettaScale Technology Inc.
+// Copyright (c) 2024 ZettaScale Technology
```
Comment (evshary) [inline src/queryable.rs:2]: I think we don't need to update the year here.

File: src/lib.rs [hunk 3]
```diff
@@ -1,5 +1,5 @@
 //
-// Copyright (c) 2017, 2022 ZettaScale Technology Inc.
+// Copyright (c) 2024 ZettaScale Technology
```
Comment (evshary) [inline src/lib.rs:2]: I think we don't need to update the year here.

File: src/config.rs [hunk 4]
```diff
@@ -1,5 +1,5 @@
 //
-// Copyright (c) 2017, 2022 ZettaScale Technology Inc.
+// Copyright (c) 2024 ZettaScale Technology
```
Comment (evshary) [inline src/config.rs:2]: I think we don't need to update the year here.

File: zenoh/__init__.py [hunk 5]
```diff
@@ -1,5 +1,5 @@
 #
-# Copyright (c) 2022 ZettaScale Technology
+# Copyright (c) 2024 ZettaScale Technology
```
Comment (evshary) [inline zenoh/__init__.py:2]: I think we don't need to update the year here.

File: src/value.rs [hunk 6]
```diff
@@ -1,431 +1,51 @@
-// Copyright (c) 2017, 2022 ZettaScale Technology Inc.
-
+//
+// Copyright (c) 2024 ZettaScale Technology
```
Comment (evshary) [inline src/value.rs:2]: I think we don't need to update the year here.

File: src/handlers.rs [hunk 7]
```diff
@@ -0,0 +1,357 @@
...
+impl FifoChannel {
+    #[new]
+    fn new(capacity: usize) -> Self {
+        Self(capacity)
+    }
+}
+
+#[pyclass]
+#[derive(Clone)]
+pub(crate) struct RingChannel(usize);
```
Comment (evshary) [inline src/handlers.rs:63]: I'm not sure if naming it as RingChannel is good or not, since Zenoh uses RingBuffer and it's not consistent.

**Review by evshary — CHANGES_REQUESTED** [review]
I guess we need to fix the conflict first.

File: src/handlers.rs (hunk 7 above)
Comment (Mallets) [reply src/handlers.rs:63]: In Rust `RingChannel` is being used. It is true that at the beginning of development it was called `RingBuffer` but now it is called `RingChannel`.

**Comment by wyfo** [comment]
@evshary I've updated the years because it has been rewritten from scratch, so I think it's pertinent.

File: src/session.rs (hunk 1 above)
Comment (wyfo) [reply src/session.rs:2]: You've a put a 👍 on my [comment](https://github.com/eclipse-zenoh/zenoh-python/pull/184#issuecomment-2068809841), so I think the thread can be resolved.

File: src/lib.rs (hunk 3 above)
Comment (wyfo) [reply src/lib.rs:2]: https://github.com/eclipse-zenoh/zenoh-python/pull/184#discussion_r1579621201

File: zenoh/__init__.py (hunk 5 above)
Comment (wyfo) [reply zenoh/__init__.py:2]: https://github.com/eclipse-zenoh/zenoh-python/pull/184#discussion_r1579621201

File: src/queryable.rs (hunk 2 above)
Comment (wyfo) [reply src/queryable.rs:2]: https://github.com/eclipse-zenoh/zenoh-python/pull/184#discussion_r1579621201

File: src/value.rs (hunk 6 above)
Comment (wyfo) [reply src/value.rs:2]: https://github.com/eclipse-zenoh/zenoh-python/pull/184#discussion_r1579621201

File: src/config.rs (hunk 4 above)
Comment (wyfo) [reply src/config.rs:4]: https://github.com/eclipse-zenoh/zenoh-python/pull/184#discussion_r1579621201

**Comment by wyfo** [comment]
Tests doesn't pass because it seems I've resolved a bug, which was taken in account in the tests. In fact, some tests was not able to be stopped with `SIGINT`, and required to be killed. 
It's no more the case with this PR. I've spent one hour trying to determine what was the cause of this bug on `main`, but I'm ashamed to admit I failed; I don't know what is causing this kind of deadlocks in the tests.
Anyway, I should update the tests to not expect programs to be killed.
@evshary If you have information about this, I would be very interested.

**Comment by wyfo** [comment]
The tests doesn't pass yet because of the deadlock issue in the examples. It's caused by relying on pseudo-RAII while exiting tests with an exception, which causes finalizers to run while Python interpreter is shutting down.
Solving this issue requires an important discussion about RAII use, so for now, the tests aren't touched.

**Comment by fuzzypixelz** [comment]
@wyfo Please change the base branch to dev/1.0.0 as protocol_changes is no longer the development branch.

**Pull request merged**